BSP/Src/BSP_I2C.c

```diff
@@ -0,0 +1,180 @@
+    uint8_t buffer[I2C_QUEUE_SIZE * I2C_ITEM_LENGTH];
+        for (uint16_t i = 0; i < len; i++) {
+            buffer[i] = pDataBuff[i];
+        }
+    if (!__HAL_I2C_GET_FLAG(hi2c, I2C_FLAG_BUSY)) {
+        HAL_I2C_Master_Transmit_IT(hi2c, deviceAdd << 1, buffer, len);
+    }
+    else if (uxQueueSpacesAvailable(I2C_DataQueue) > len) {
+        for (uint32_t i = 0; i < len; i++) {
+            xQueueSend(I2C_DataQueue, buffer[i], 0);
+        }
+            // HAL requires 7b I2C address left shifted
+            xQueueSend(I2C_MetaQueue, deviceAdd << 1, 0);
+            xQueueSend(I2C_MetaQueue, len, 0);
+    }
+}
+
+//overwrite this
+void HAL_I2C_MasterTxCpltCallback(I2C_HandleTypeDef *hi2c) {
+    uint8_t buffer[I2C_QUEUE_SIZE * I2C_ITEM_LENGTH];
+    if (xQueueIsQueueEmptyFromISR(I2C_MetaQueue) == pdFALSE) {
+        uint32_t deviceAdd, len;
+        uint8_t temp;
+        xQueueReceiveFromISR(I2C_MetaQueue, &deviceAdd, &temp);
+        xQueueReceiveFromISR(I2C_MetaQueue, &len, &temp);
+        for (uint32_t i = 0; i < len; i++) {
+            xQueueReceiveFromISR(I2C_DataQueue, &(buffer[i]), &temp);
+        }
+        HAL_I2C_Master_Transmit_IT(hi2c, deviceAdd, buffer, len);
+    }
+}
+
+/** 
+* @param    hi2c:            to structure that has config data
+* @param    deviceAdd :      target device address
+* @param    memoryAdd :      
+* @param    memoryAddSize :  
+* @param    pDataBuff :      data buffer
+* @param    len :            amount of data   
+* @return   HAL Status    
+*/
+void I2CRegisterWrite(I2C_HandleTypeDef *hi2c, 
+              uint32_t deviceAdd, 
+              uint32_t memoryAdd, //register
+              uint32_t memoryAddSize,
+              uint8_t* pDataBuff, 
+              uint16_t len) {
+    //copy buffer data over for security incase it changes 
+    uint8_t buffer[I2C_QUEUE_SIZE * I2C_ITEM_LENGTH];
+        for (uint16_t i = 0; i < len; i++) {
+            buffer[i] = pDataBuff[i];
+        }
+    if (!__HAL_I2C_GET_FLAG(hi2c, I2C_FLAG_BUSY)) {
+        HAL_I2C_Mem_Write_IT(hi2c, deviceAdd << 1, memoryAdd, memoryAddSize, buffer, len);
+    }
+    else if (uxQueueSpacesAvailable(I2C_DataQueue) > len) {
+        for (uint32_t i = 0; i < len; i++) {
+            xQueueSend(I2C_DataQueue, buffer[i], 0);
+        }
+            // HAL requires 7b I2C address left shifted
+            xQueueSend(I2C_MetaQueue, deviceAdd << 1, 0);
+            xQueueSend(I2C_MetaQueue, memoryAdd, 0);
+            xQueueSend(I2C_MetaQueue, memoryAddSize, 0);
+            xQueueSend(I2C_MetaQueue, len, 0);
+    }
+}
+
+//register overwrite version
+void HAL_I2C_MemTxCpltCallback(I2C_HandleTypeDef *hi2c) {
+    uint8_t buffer[I2C_QUEUE_SIZE * I2C_ITEM_LENGTH];
+    if (xQueueIsQueueEmptyFromISR(I2C_MetaQueue) == pdFALSE) {
+        uint32_t deviceAdd, memoryAdd, memoryAddSize, len;
+        uint8_t temp;
+        xQueueReceiveFromISR(I2C_MetaQueue, &deviceAdd, &temp);
```

Comment:
should be good because it's a queue but need to make sure data pop order is correct